services/rpcfilters/hash_filter.go

```diff
@@ -0,0 +1,56 @@
+package rpcfilters
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type hashFilter struct {
+	hashes []common.Hash
+	mu     sync.Mutex
+	done   chan struct{}
+	timer  *time.Timer
+}
+
+// add adds a hash to the hashFilter
+func (f *hashFilter) add(data interface{}) error {
+	hash, ok := data.(common.Hash)
+	if !ok {
+		return errors.New("provided data is not a common.Hash")
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.hashes = append(f.hashes, hash)
+	return nil
+}
+
+// pop returns all the hashes stored in the hashFilter and clears the hashFilter contents
+func (f *hashFilter) pop() interface{} {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	hashes := f.hashes
+	f.hashes = nil
+	return returnHashes(hashes)
```

Comment:
`returnHashes` function probably should be moved into this file as well, for locality.